### PR TITLE
Only log warning when actually failing shards

### PR DESCRIPTION
**Pull request by dnhatn**
Currently the master node logs a warning message whenever it receives a failed shard request. However, this can be noisy because
- Multiple failed shard requests can be issued for a single shard
- Failed shard requests can be still issued for an already failed shard

This commit moves the log-warn to AllocationService in which the failing shard action actually happens. This is another prerequisite step in order to not ignore the shard not-available exceptions in the replication.

Relates #28534

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/cluster/routing/allocation/AllocationService.java [hunk 1]
```diff
@@ -188,6 +189,7 @@ public ClusterState applyFailedShards(final ClusterState clusterState, final Lis
                 if (failedShardEntry.markAsStale()) {
                     allocation.removeAllocationId(failedShard);
                 }
+                logger.warn(new ParameterizedMessage("failing shard [{}]", failedShardEntry), failedShardEntry.getFailure());
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/cluster/routing/allocation/AllocationService.java]: also add logging information for the "mark as stale" part here as well as in removeStaleIdsWithoutRoutings?

**Comment by dnhatn** [comment]
@ywelsch, I've added a logging for `stale shards`. The `failedShardEntry` already includes the `mark as stale` information. Please have another look. Thank you!

File: server/src/main/java/org/elasticsearch/cluster/routing/allocation/IndexMetaDataUpdater.java [hunk 2]
```diff
@@ -238,6 +239,11 @@ public static ClusterState removeStaleIdsWithoutRoutings(ClusterState clusterSta
                     }
                     indexMetaDataBuilder.putInSyncAllocationIds(shardNumber, remainingInSyncAllocations);
                 }
+                // Only log the stale shards which have been actually removed.
+                if (oldInSyncAllocations.size() != remainingInSyncAllocations.size()) {
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/cluster/routing/allocation/IndexMetaDataUpdater.java]: This check is not necessary. `idsToRemove` already contains the info that's needed (it does not contain non-existing ids, you can even add an assertion for that if you like).
Also I would change the message to "{} marking unavailable shards as stale: {}"

Comment (dnhatn) [reply server/src/main/java/org/elasticsearch/cluster/routing/allocation/IndexMetaDataUpdater.java]: I pushed https://github.com/elastic/elasticsearch/pull/28558/commits/b0cd95476487d89efd5d28799e00649a511a6be1

**Review by ywelsch — APPROVED** [review]
LGTM

**Comment by dnhatn** [comment]
Thanks @ywelsch for reviewing.

**Pull request merged**